Fix ApplyUnary -(1.s) and -(-a)

MasseGuillaume opened the pull request.

MasseGuillaume committed:
Fix ApplyUnary -(1.s) and -(-a)

MasseGuillaume commented on tests/slow/coverage.txt:
```diff
@@ -10,7 +10,6 @@ scala/test/files/pos/t9050.scala
 scala/test/files/run/t2378.scala
 scala/test/files/run/t6506.scala
 scala/test/files/specialized/arrays.scala
-scalafx/scalafx/src/test/scala/scalafx/util/DurationSpec.scala
```
15 to go

olafurpg commented on format/src/main/scala/scala/meta/internal/fmt/SyntacticGroup.scala:
```diff
@@ -50,14 +52,8 @@ object SyntacticGroup {
       Literal.precedence == Pat.SimplePattern.precedence
   )
-  case object Path extends Type with Term with Pat {
+  case class Path(tree: Tree) extends Type with Term with Pat {
```
Would it be possible to avoid this tree field by handling the needs parentheses case in ApplyUnary? I'm a bit concerned about taxing all usages of Path when we could handle the numeric literal case at the root.

MasseGuillaume replied:
You prefer to have all the parenthesis logic in TreePrinter ?

olafurpg replied:
No 😉 but in this particular case for unary apply on numeric literal however it seems custom handling in TreePrinter beats introducing complexity into syntactic groups, which is used in a lot of places. It's cleaner to see in TreePrinter that unary apply requires custom handling than make syntactic groups look more complicated than they already are.

olafurpg commented on tests/unit/src/test/scala/org/scalafmt/tests/TermApplyUnarySuite.scala:
```diff
@@ -5,4 +5,9 @@ object TermApplyUnarySuite extends BaseScalaPrinterTest {
   check("!a.b")
   check("!(if (a) b else c)")
   check("!(a op b)")
+  check("-(1.s)")
```
💯

olafurpg commented on format/src/main/scala/org/scalafmt/internal/TreePrinter.scala:
```diff
@@ -250,7 +250,8 @@ object TreePrinter {
             print(t.expr) + space + `wildcard`
           case t: Term.ApplyUnary =>
-            print(t.op) + SimpleExpr.wrap(t.arg)
+            val group = TreeSyntacticGroup(t)
```
Can we do `PrefixExpr(t.op.value)` instead? I think that would be consistent with how it's done in other cases

MasseGuillaume replied:
I would prefer to go the other way around: add more of those.

olafurpg replied:
Will this every be anything else than `PrefixExpr(t.op.value)`?

MasseGuillaume replied:
We already have a function to compute this. Let's not duplicate stuff.

olafurpg replied:
I agree it would make sense to consolidate TreeSyntacticGroups with the custom SimpleExpr/... in TreePrinter.

Can we use `PrefixExpr(t.op.value)` here for consistency with the rest of TreePrinter and then follow up in another PR to migrate everything to TreeSyntacticGroups (which we might want to rename)? Consistency makes the code easier to follow, when I saw this change here I thought something special was happening that required extra attention.

olafurpg commented:
I think it's fine to make the ApplyUnary case more complicated if that avoids leaking complexity into syntactic groups.

MasseGuillaume committed:
Add non-lit case for UnaryApply

MasseGuillaume committed:
new SyntacticGroup PrefixArg to capture PrefixExpr rhs

For Example
-(1.s)

We wrap 1.s in PrefixArg. We use the tree part to match for Lit

We also keep the SyntacticGroup to match cases like -(-a)
